### PR TITLE
Add the ability to listen for replies to specific messages

### PR DESCRIPTION
**Pull request by Mysterypotatoguy**
## Checklist
- [x] I have tested this PR[^1].
- [x] I have read the [contributing guidelines](./CONTRIBUTING.md).

## Changelog
- Adds message reply listeners (via `Message#addMessageCreateListener`)

### Breaking Changes
N/A

## Description

Adds the ability to attach message create listeners to messages, which will be fired when the message receives a reply 

Closes: #1114

[^1]: At least started a running bot instance with your changes and triggered an event so your changed code gets executed.

### CODE REVIEW

**Review by KILLEliteMaste — COMMENTED** [review]
Imo the `MessageCreateListener` doesn't fit name wise for this purpose. We are not listening for MessageCreates but rather for MessageReplies. And adding a `MessageCreateListener` to a message is not that obvious that you are waiting for replies. For that reason I think a new `MessageReplyListener` (can extend from `MessageCreateListener`) would make more sense

**Comment by KILLEliteMaste** [comment]
> Should a `MessageReplyListener` have its own event `MessageReplyEvent` or should it still use the `MessageCreateEvent`?

I think it's own event is better. Then we can also add `getReferencedMessage` and `requestReferencedMessage` to the event like we have for the message content.

**Comment by Saladoc** [comment]
Semantics question for `MessageReplyEvent`: 
If attached to a `User` / `MessageAuthor`, does it 
 * Trigger for messages in Reply to that user / author?
 * Trigger for messages by that user / author?

**Comment by Mysterypotatoguy** [comment]
It doesn't necessarily have to be attachable to `User`/`MessageAuthor`

**Comment by Vampire** [comment]
Why not both?

File: javacord-api/src/main/java/org/javacord/api/listener/message/MessageReplyListener.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import org.javacord.api.event.message.MessageReplyEvent;
+import org.javacord.api.listener.ObjectAttachableListener;
+import org.javacord.api.listener.user.UserAttachableListener;
+
+/**
+ * This listener listens to message replies.
+ */
+@FunctionalInterface
+public interface MessageReplyListener extends UserAttachableListener, MessageAttachableListener,
+        ObjectAttachableListener {
```
Comment (KILLEliteMaste) [inline javacord-api/src/main/java/org/javacord/api/listener/message/MessageReplyListener.java:12]: Shouldn't the interface extend the same interfaces as the `MessagecreateListener`?

File: javacord-core/src/main/java/org/javacord/core/event/message/MessageReplyEventImpl.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+import org.javacord.api.entity.message.Message;
+import org.javacord.api.event.message.MessageReplyEvent;
+
+/**
+ * The implementation of {@link MessageReplyEvent}.
+ */
+public class MessageReplyEventImpl extends CertainMessageEventImpl implements MessageReplyEvent {
+
+    private Message referencedMessage;
```
Comment (KILLEliteMaste) [inline javacord-core/src/main/java/org/javacord/core/event/message/MessageReplyEventImpl.java:11]:
```suggestion
    private final Message referencedMessage;
```

File: javacord-core/src/main/java/org/javacord/core/util/handler/message/MessageCreateHandler.java [hunk 3]
```diff
@@ -98,6 +105,20 @@ private void handle(TextChannel channel, JsonNode packet) {
                 author.asUser().orElse(null),
                 author.isWebhook() ? author.getId() : null,
                 event);
+
+        optionalReferencedMessage.ifPresent(referencedMessage -> {
```
Comment (KILLEliteMaste) [inline javacord-core/src/main/java/org/javacord/core/util/handler/message/MessageCreateHandler.java:109]:
```suggestion
        message.getReferencedMessage().ifPresent(referencedMessage -> {
```
And remove the variable assignment in line 93

**Review by KILLEliteMaste — CHANGES_REQUESTED** [review]

File: javacord-api/src/main/java/org/javacord/api/listener/message/MessageReplyListener.java (hunk 1 above)
Comment (Mysterypotatoguy) [reply javacord-api/src/main/java/org/javacord/api/listener/message/MessageReplyListener.java:12]: Not sure, do we really want this being attachable to Servers and TextChannels? The semantics seem very weird to me

Comment (Vampire) [reply javacord-api/src/main/java/org/javacord/api/listener/message/MessageReplyListener.java:12]: Why not? You listen to replies happening in a specific channel or server. I cannot see where that would be weird.

File: javacord-core/src/main/java/org/javacord/core/util/handler/message/MessageCreateHandler.java [hunk 4]
```diff
@@ -98,6 +104,31 @@ private void handle(TextChannel channel, JsonNode packet) {
...
+            referencedMessage.getUserAuthor().ifPresent(users::add);
+            message.getUserAuthor().ifPresent(users::add);
+
+            Set<Long> webhookIds =  new HashSet<>();
+            if (message.getAuthor().isWebhook()) {
+                webhookIds.add(message.getAuthor().getId());
+            }
+            if (referencedMessage.getAuthor().isWebhook()) {
+                webhookIds.add(referencedMessage.getAuthor().getId());
+            }
```
Comment (KILLEliteMaste) [inline javacord-core/src/main/java/org/javacord/core/util/handler/message/MessageCreateHandler.java:120]:
```suggestion
            message.getAuthor().getWebhookId().ifPresent(webhookIds::add);
            referencedMessage.getAuthor().getWebhookId().ifPresent(webhookIds::add);
```

**Review by KILLEliteMaste — CHANGES_REQUESTED** [review]

**Review by KILLEliteMaste — APPROVED** [review]
LGTM!

**Pull request merged**